### PR TITLE
Indexing queue page

### PR DESCRIPTION
**Pull request by damcou**
Add dedicated indexing queue page feature

### CODE REVIEW

File: Model/Job.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+class Job extends \Magento\Framework\Model\AbstractModel implements IdentityInterface, JobInterface
+{
+    const CACHE_TAG = 'algoliasearch_queue_job';
+
+    protected $_cacheTag = 'algoliasearch_queue_job';
+
+    protected $_eventPrefix = 'algoliasearch_queue_job';
+
+    /**
+     * Object Manager
```
Comment (JanPetr) [inline Model/Job.php]: Not needed

File: Model/Job.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        if ((int) $this->getRetries() >= $this->getMaxRetries()) {
+            $status = JobInterface::STATUS_ERROR;
+        }
+
+        return $status;
+    }
+
+    /**
+     * execute the job
```
Comment (JanPetr) [inline Model/Job.php]: Not needed

File: Model/Job.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        $this->setRetries((int) $this->getRetries() + 1);
+        call_user_func_array([$model, $method], $data);
+        $this->getResource()->save($this);
+
+        return $this;
+    }
+
+    /**
+     * Save the error
```
Comment (JanPetr) [inline Model/Job.php]: Not needed

File: Model/Job.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
+<?php
+namespace Algolia\AlgoliaSearch\Model;
```
Comment (JanPetr) [inline Model/Job.php]: Can you leave an empty line between `<?php` and `namespace`? Applies for all the files

File: Model/ResourceModel/Job.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
+<?php
+namespace Algolia\AlgoliaSearch\Model\ResourceModel;
+
+
```
Comment (JanPetr) [inline Model/ResourceModel/Job.php]: Leave just one line between `namespace` and `class`

File: Model/ResourceModel/Job.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+class Job extends \Magento\Framework\Model\ResourceModel\Db\AbstractDb
+{
+
+    public function __construct(
+        \Magento\Framework\Model\ResourceModel\Db\Context $context
```
Comment (JanPetr) [inline Model/ResourceModel/Job.php]: Can you add `\Magento\Framework\Model\ResourceModel\Db\Context` to `use` and one-line it?

File: Model/ResourceModel/Job.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        parent::__construct($context);
+    }
+
+    protected function _construct()
+    {
+        $this->_init('algoliasearch_queue', 'job_id');
+    }
+
+    /**
+     * delete a list of entities by the ids
```
Comment (JanPetr) [inline Model/ResourceModel/Job.php]: Not needed

File: Model/ResourceModel/Job.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+     *
+     * @param int[] $ids ids to delete
```
Comment (JanPetr) [inline Model/ResourceModel/Job.php]: `ids` => `IDs`

File: Model/ResourceModel/Job.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    public function deleteIds($ids)
+    {
+        $condition = $this->getConnection()->quoteInto($this->getIdFieldName() . ' IN (?)', (array) $ids);
+        $this->getConnection()->delete($this->getMainTable(), $condition);
+
+        return $this;
+    }
+
+    /**
+     * delete all entities
```
Comment (JanPetr) [inline Model/ResourceModel/Job.php]: Not needed

File: Model/ResourceModel/Job/Grid/Collection.php [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+     * @param AggregationInterface $aggregations
+     * @return $this
+     */
+    public function setAggregations($aggregations)
+    {
+        $this->aggregations = $aggregations;
+    }
+
+    /**
+     * Get search criteria.
```
Comment (JanPetr) [inline Model/ResourceModel/Job/Grid/Collection.php]: Not needed

File: Model/ResourceModel/Job/Grid/Collection.php [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+     *
+     * @return \Magento\Framework\Api\SearchCriteriaInterface|null
+     */
+    public function getSearchCriteria()
+    {
+        return null;
+    }
+
+    /**
+     * Set search criteria.
```
Comment (JanPetr) [inline Model/ResourceModel/Job/Grid/Collection.php]: Not needed

File: Model/ResourceModel/Job/Grid/Collection.php [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+     * @return $this
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function setSearchCriteria(\Magento\Framework\Api\SearchCriteriaInterface $searchCriteria = null)
+    {
+        return $this;
+    }
+
+    /**
+     * Get total count.
```
Comment (JanPetr) [inline Model/ResourceModel/Job/Grid/Collection.php]: Not needed

File: Model/ResourceModel/Job/Grid/Collection.php [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+     *
+     * @return int
+     */
+    public function getTotalCount()
+    {
+        return $this->getSize();
+    }
+
+    /**
+     * Set total count.
```
Comment (JanPetr) [inline Model/ResourceModel/Job/Grid/Collection.php]: Not needed

File: Model/ResourceModel/Job/Grid/Collection.php [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+     *
+     * @param int $totalCount
+     * @return $this
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function setTotalCount($totalCount)
+    {
+        return $this;
```
Comment (JanPetr) [inline Model/ResourceModel/Job/Grid/Collection.php]: The method doesn't do anything, it just returns $this. Is it a bug? If not, is the method needed at all?

File: Model/ResourceModel/Job/Grid/Collection.php [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Set search criteria.
+     *
+     * @param \Magento\Framework\Api\SearchCriteriaInterface $searchCriteria
+     * @return $this
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function setSearchCriteria(\Magento\Framework\Api\SearchCriteriaInterface $searchCriteria = null)
+    {
+        return $this;
```
Comment (JanPetr) [inline Model/ResourceModel/Job/Grid/Collection.php]: The method doesn't do anything, it just returns $this. Is it a bug? If not, is the method needed at all?

File: Model/ResourceModel/Job/Grid/Collection.php [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Set items list.
+     *
+     * @param \Magento\Framework\Api\ExtensibleDataInterface[] $items
+     * @return $this
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function setItems(array $items = null)
+    {
+        return $this;
```
Comment (JanPetr) [inline Model/ResourceModel/Job/Grid/Collection.php]: The method doesn't do anything, it just returns $this. Is it a bug? If not, is the method needed at all?

File: Ui/Component/Listing/Column/Data.php [hunk 17]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\Ui\Component\Listing\Column;
+
+class Data extends \Magento\Ui\Component\Listing\Columns\Column
+{
+    /**
+     * Prepare Data Source
```
Comment (JanPetr) [inline Ui/Component/Listing/Column/Data.php]: Not needed

File: Ui/Component/Listing/Column/QueueActions.php [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+namespace Algolia\AlgoliaSearch\Ui\Component\Listing\Column;
+
+use Magento\Framework\UrlInterface;
+use Magento\Framework\View\Element\UiComponent\ContextInterface;
+use Magento\Framework\View\Element\UiComponentFactory;
+use Magento\Ui\Component\Listing\Columns\Column;
+use Magento\Framework\Escaper;
+
+/**
+ * Class QueueActions
```
Comment (JanPetr) [inline Ui/Component/Listing/Column/QueueActions.php]: Not needed

File: Ui/Component/Listing/Column/QueueActions.php [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+ */
+class QueueActions extends Column
+{
+    /**
+     * Url path
```
Comment (JanPetr) [inline Ui/Component/Listing/Column/QueueActions.php]: Not needed

File: Ui/Component/Listing/Column/QueueActions.php [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    const URL_PATH_VIEW = 'algolia_algoliasearch/queue/view';
+
+    /**
+     * @var UrlInterface
+     */
+    protected $urlBuilder;
+
+    /**
+     * Constructor
```
Comment (JanPetr) [inline Ui/Component/Listing/Column/QueueActions.php]: Not needed

File: Ui/Component/Listing/Column/QueueActions.php [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+        array $components = [],
+        array $data = []
+    ) {
+        $this->urlBuilder = $urlBuilder;
+
+        parent::__construct($context, $uiComponentFactory, $components, $data);
+    }
+
+    /**
+     * Prepare Data Source
```
Comment (JanPetr) [inline Ui/Component/Listing/Column/QueueActions.php]: Not needed

File: Ui/Component/Listing/Column/QueueActions.php [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+     *
+     * @param array $dataSource
+     * @return array
+     */
+    public function prepareDataSource(array $dataSource)
+    {
+        if (isset($dataSource['data']['items'])) {
+            foreach ($dataSource['data']['items'] as &$item) {
+                $item[$this->getData('name')] = [
+//                    'delete' => [
```
Comment (JanPetr) [inline Ui/Component/Listing/Column/QueueActions.php]: Can you delete the code completely if it's not used at all?

File: Ui/Component/Listing/Column/Status.php [hunk 23]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\Ui\Component\Listing\Column;
+
+use Algolia\AlgoliaSearch\Api\Data\JobInterface;
+
+class Status extends \Magento\Ui\Component\Listing\Columns\Column
+{
+    /**
+     * Prepare Data Source
```
Comment (JanPetr) [inline Ui/Component/Listing/Column/Status.php]: Not needed

File: Ui/Component/Listing/Column/Status.php [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+        foreach ($dataSource['data']['items'] as &$item) {
+            $item[$fieldName] = $this->defineStatus($item);
+        }
+
+        return $dataSource;
+    }
+
+
+    /**
+     * Define the job status
```
Comment (JanPetr) [inline Ui/Component/Listing/Column/Status.php]: Not needed

File: Controller/Adminhtml/Queue/Clear.php [hunk 25]
```diff
@@ -3,22 +3,23 @@
...
     {
+        /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
+        $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+        $resultRedirect->setPath('*/*/index');
+
         try {
             $this->db->query('TRUNCATE TABLE '.$this->tableName);
-
-            $status = ['status' => 'ok'];
+            $this->messageManager->addNoticeMessage(__('Queue has been cleared'));
```
Comment (JanPetr) [inline Controller/Adminhtml/Queue/Clear.php]: Add `.` on the end of the sentence.

File: Controller/Adminhtml/Queue/Index.php [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+        $resultPage->getConfig()->getTitle()->prepend($breadMain);
+
+        return $resultPage;
+    }
+
+    protected function checkQueueIsActivated()
+    {
+        if (! $this->configHelper->isQueueActive()) {
+            $msg = __(
+                "The indexing queue is not activated. Please activate it in the <a href=\"%1\">Algolia configuration</a>",
```
Comment (JanPetr) [inline Controller/Adminhtml/Queue/Index.php]: Add `.` on the end of the sentence.

File: Model/ResourceModel/Job/Grid/Collection.php [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+namespace Algolia\AlgoliaSearch\Model\ResourceModel\Job\Grid;
+
+use Magento\Framework\Api\Search\SearchResultInterface;
+use Magento\Framework\Api\Search\AggregationInterface;
+use Algolia\AlgoliaSearch\Model\ResourceModel\Job\Collection as JobCollection;
+
+class Collection extends JobCollection implements SearchResultInterface
+{
+    /**
+     * @var AggregationInterface
```
Comment (JanPetr) [inline Model/ResourceModel/Job/Grid/Collection.php]: One-line it :)

File: Model/ResourceModel/Job/Grid/Collection.php (hunk 15 above)
Comment (damcou) [reply Model/ResourceModel/Job/Grid/Collection.php]: It's coming from SearchResultsInterface , and it has to be implemented.

File: Model/ResourceModel/Job/Grid/Collection.php (hunk 14 above)
Comment (damcou) [reply Model/ResourceModel/Job/Grid/Collection.php]: It's coming from SearchResultsInterface , and it has to be implemented.

File: Model/ResourceModel/Job/Grid/Collection.php (hunk 16 above)
Comment (damcou) [reply Model/ResourceModel/Job/Grid/Collection.php]: It's coming from SearchResultsInterface , and it has to be implemented.

File: Block/Adminhtml/Queue/Status.php [hunk 28]
```diff
@@ -0,0 +1,178 @@
...
+    {
+        $notices = [];
+
+        if ($this->isQueueStuck()) {
+            $notices[] = '<a href="' . $this->getResetQueueUrl() .'"> ' . __('Reset queue') . '</a>';
+        }
+
+        if ($this->isQueueNotProcessed()) {
+            $notices[] =  __(
+                'We noticed that the queue has not been processed for one hour. Can you please check its status and if you correctly set up the cron?'
```
[the remark this answers is not in the document]
Comment (JanPetr) [reply Block/Adminhtml/Queue/Status.php]: Applies to all suggestions - stuck, slow, no processing queue

File: Model/ResourceModel/Job.php [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        $this->_init('algoliasearch_queue', 'job_id');
+    }
+
+    /**
+     * @param int[] $ids IDs to delete
+     *
+     * @return Job
+     */
+    public function deleteIds($ids)
```
Comment (JanPetr) [inline Model/ResourceModel/Job.php]: As we are not deleting single jobs, do we need this method?

File: Block/Adminhtml/Queue/Status.php (hunk 28 above)
[the remark this answers is not in the document]
Comment (JanPetr) [reply Block/Adminhtml/Queue/Status.php]: And I have a feeling we might work on the exact wording, but I'm not the best one on it. Let's discuss during the sync :)

**Pull request merged**